Raycast attach node to check for colliders

taniwha opened the pull request:
The ray cast is used to check whether then attach node is inside a
collider on the part to be dropped, and if so find the offset needed to
avoid dropping the part with the collider inside another collider, thus
preventing the part from flying off in low gravity. Of course, the
offset will not be sufficient of the collider surface is not
perpendicular to the attach node, or if the overall part collider setup
is concave around the attach node.

The computed offset works as a default value for the user-adjustable
offset.

taniwha committed:
Raycast attach node to check for colliders

The ray cast is used to check whether then attach node is inside a
collider on the part to be dropped, and if so find the offset needed to
avoid dropping the part with the collider inside another collider, thus
preventing the part from flying off in low gravity. Of course, the
offset will not be sufficient of the collider surface is not
perpendicular to the attach node, or if the overall part collider setup
is concave around the attach node.

The computed offset works as a default value for the user-adjustable
offset.

ihsoft commented on Source/api/Utils/PartUtilsImpl.cs:
```diff
@@ -143,7 +143,7 @@ public class PartUtilsImpl {
         continue;  // They must be handled before the connected RBs handled.
       }
-      if (!(component is Renderer || component is MeshFilter)) {
+      if (!(component is Renderer || component is MeshFilter || component is Collider)) {
```
This changes the method behavior which is declared in the description. The returned model is expected to be physics free, and with colliders it's not. Could you please add an optional parameter to keep the colliders and update the description? Another way could be switching the colliders to triggers, and explaining this behavior in the method description.

ihsoft commented on Source/KISAddonPointer.cs:
```diff
@@ -750,6 +784,12 @@ public enum PointerState {
     // Collect models from all the part in the assembly.
     pointer = new GameObject("KISPointer");
     var model = KISAPI.PartUtils.GetSceneAssemblyModel(rootPart);
+    var colliders = KISAPI.PartUtils.FindColliders (model);
```
You don't need a new API method for this:
```
var colliders = model.GetComponentsInChildren<Collider>().ToList();
```

ihsoft commented on Source/api/Utils/PartUtilsImpl.cs:
```diff
@@ -169,6 +169,18 @@ public class PartUtilsImpl {
     return modelObj;
   }
 
+  public List<Collider> FindColliders (GameObject modelObj)
```
This functionality can be achieved via a simple one-liner. I don't think KIS API needs such a wrapper.

ihsoft commented on Source/KISAddonPointer.cs:
```diff
@@ -720,6 +726,34 @@ public enum PointerState {
+
+  static void CreateAutoOffsets(List<Collider> colliders) {
+    float distance = 1;
+    int layerMask = 1 << 0;
```
```
var layerMask = KspLayerMask.Part;
```

How about checking for the surface colliders too?

ihsoft replied:
FYI on the masks: https://ihsoft.github.io/KSPDev_Utils/v2.3/html/T_KSPDev_ModelUtils_KspLayerMask.htm

taniwha replied:
Not desired. The idea is to check for the colliders on the part being placed, not the target. Checking for surface colliders could cause more false positives that would need to be filtered against the colliders in the part.

The ReStock Mk1-3 capsule has the bottom attach node 20cm inside the collider. Thus, when the capsule is placed on the ground, it is thrown upwards quite violently.

ihsoft replied:
I used to drop `dockingPort3` on the Minmus surface, and the output was always an explosion. And the reason was exactly this: an attach node inside the part's collider. Anyway, it's up to you here. I gave up to fix in KIS v1 anyway.

taniwha replied:
dockingPort3... I'll take a look (I haven't looked into the annoying small parts).

taniwha committed:
Address PR feedback

ihsoft commented on Source/KISAddonPointer.cs:
```diff
@@ -783,8 +783,8 @@ static int CompareDistance(RaycastHit a, RaycastHit b)
     pointer = new GameObject("KISPointer");
-    var model = KISAPI.PartUtils.GetSceneAssemblyModel(rootPart);
-    var colliders = KISAPI.PartUtils.FindColliders (model);
+    var model = KISAPI.PartUtils.GetSceneAssemblyModel(rootPart, true, true);
```
Nit:
```
var model = KISAPI.PartUtils.GetSceneAssemblyModel(rootPart, keepColliders: true);
```

ihsoft commented on Source/api/Utils/PartUtilsImpl.cs:
```diff
@@ -113,7 +113,7 @@ public class PartUtilsImpl {
   /// needed anymore.
   /// </returns>
-  public GameObject GetSceneAssemblyModel(Part rootPart, bool goThruChildren = true) {
+  public GameObject GetSceneAssemblyModel(Part rootPart, bool goThruChildren = true, bool keepColliders = false) {
```
Even though the name of the parameter is self-explanatory, could you please modify the doc too? It helps running doc generation tools.

taniwha replied:
Ah, sorry, I forgot to do so.

ihsoft reviewed (commented):
How about using `KspLayerMask.Part` instead of magic constant `1 << 0`?

taniwha commented:
I've taken care of those issues but not pushed yet. Last night, I noticed that dropping RCS ports was behaving strangely (offset probably 1m from the surface, but attaching worked properly). I suspect the ray cast is failing to hit anything and thus the offset is getting set to 1 instead of 0 (though I think RCS ports have trouble with being dropped, so I need to investigate further than just that). I'll push an update as soon as I can.

taniwha committed:
Fix more PR feedback issues

taniwha committed:
Transform node position and direction

This fixes the weird offsets when dropping certain parts.

taniwha commented:
I have also found out why small parts explode when placed on the ground. It's actually due to the _lack_ of colliders. Ie, the part origin (and thus vessel origin since part origin -> part CoM (default) -> vessel CoM) is outside of any collider and thus can pass into the ground. Vessel does not like this. Placing the part upside down works around this, but is annoying. I am investigating a potential fix.

taniwha commented:
Actually, that transform patch worked only on the launch pad, not out in the wild. Still working on it.

ihsoft commented:
Please, let me know when it's OK to merge.

taniwha commented:
Yeah, I'll do so. I hope to have something soon.

taniwha committed:
Cope with the inconsistency of surface attach nodes

It seems the in/out direction of the surface attach node makes no
difference to part orientation during editor or field placement (must be
some sort of cancellation), thus there's no consistency between parts
(even stock) because there was nothing to enforce it. Therefore, it's
necessary to test both directions. The smaller of the two offsets is
used on the assumption that the node is at least positioned correctly.

taniwha commented:
@ihsoft All done. That surface attach node issue was a nightmare.
